Feature/tags for projects

adrianz-kainos opened the pull request:
Resolves #21 
To manage tags, go to a project radar page e.g. `/project/1` and click 'Tags' on the black navigation bar. Tags appear in a separate column in the `/projects` search page.

adrianz-kainos committed:
Add tags tables and API

adrianz-kainos committed:
Add the ability to manage tags

adrianz-kainos committed:
Add missing anchor

xjedam commented on dao/tag.js:
```diff
@@ -0,0 +1,166 @@
+};
+
+/**
+ * Delete a set of links of tags with projects
```
misleading comment

adrianz-kainos replied:
Fixed.

xjedam commented on dao/tag.js:
```diff
@@ -0,0 +1,166 @@
+    for (var i = 2; i <= tagIds.length+1; i++) {
+        // gives us (tagId, projectId) placeholders to insert after the VALUES statement
+        sql += ` ($` + i + `, $1)` 
+        if (i != tagIds.length+1) { 
```
a lot of loopo-ifology, have you considered using .map() and/or .join() instead?

adrianz-kainos replied:
Fixed. thx

xjedam commented on dao/tag.js:
```diff
@@ -0,0 +1,166 @@
+ * @param done Function to call with the results
+ */
+Tag.getAll = function (done) {
+    var sql = `SELECT * FROM tags`
```
Could we use dbhelper.getAllFromTable here?

adrianz-kainos replied:
Yup, makes it nice and short.

xjedam commented on handlers/api/tagsApiHandler.js:
```diff
@@ -0,0 +1,92 @@
+
+TagsApiHandler.getTags = function (req, res) {
+    tag.getAll(function (result) {
+        res.writeHead(200, { "Content-Type": "application/json" });
```
perhaps we could standarise and simplify by using ApiUtils.handleResultSet ?

adrianz-kainos replied:
Done.

xjedam commented on handlers/api/tagsApiHandler.js:
```diff
@@ -0,0 +1,92 @@
+TagsApiHandler.getForProject = function (req, res) {
+    var projectId = sanitizer(req.params.projectId);
+    tag.getForProject(projectId, function (result) {
+        res.writeHead(200, { "Content-Type": "application/json" });
```
like above, we could use handleResultSet - this also covers for returning errors.

adrianz-kainos replied:
Done.

xjedam commented on views/pages/attachTags.ejs:
```diff
@@ -0,0 +1,213 @@
+<script src="/lib/bootstrap-table/bootstrap-table.min.js"></script>
+<script>
+    $table = $('#results');
+    $deleteBtn = $('#deleteBtn')
```
';' missing

adrianz-kainos replied:
Added.

xjedam commented on views/pages/attachTags.ejs:
```diff
@@ -0,0 +1,213 @@
+
+</script>
+
+<script>
```
Any reason this is a separate script tag?

adrianz-kainos replied:
Nope, removed it. Thx.

xjedam commented on views/pages/attachTags.ejs:
```diff
@@ -0,0 +1,213 @@
+            "Note that the name will change for all projects that already use this tag.", 
+            getNameSelections()[0]);
+
+        if(data.name == null){
```
Perhaps we could do some tag name validation? At least one latter perhaps - what do you think?

adrianz-kainos replied:
Done.

xjedam commented on views/pages/attachTags.ejs:
```diff
@@ -0,0 +1,213 @@
+
+<script>
+    function addTag() {
+        var tagName = prompt("Please enter the name of a new tag.");
```
Maybe some validation here as well?

adrianz-kainos replied:
Done.

xjedam commented on views/pages/attachTags.ejs:
```diff
@@ -0,0 +1,213 @@
+            if(msg.success) {
+                location.reload();
+            } else {
+                alert("Unable to add this tag. Perhaps the same tag already exists.");
```
we currently validate this only server-side, but just a generic error is shown here. There are other places this happens as well.

adrianz-kainos replied:
Done. A shared validator has been added.

xjedam commented on handlers/api/tagsApiHandler.js:
```diff
@@ -0,0 +1,92 @@
+TagsApiHandler.updateTag = function (app) {
+    return function (req, res) {
+        var tagId = sanitizer(req.body.tag);
+        var name = sanitizer(req.body.name);
```
might be worth adding name validation. Have you considered using shared validators - or thought of some other mechanism for not replicating the same validation across multiple handlers/templates?

adrianz-kainos replied:
Shared tag validator has been added.

xjedam commented on views/pages/attachTags.ejs:
```diff
@@ -0,0 +1,213 @@
+                <div class="form-inline" role="form">
+                    <button id="submitBtn" href="#" class="btn btn-success">
+                        <i class="glyphicon glyphicon-ok"></i>
+                        Attach selected
```
Any particular reason Attaching a tag is replicated across two screens?

adrianz-kainos replied:
@xjedam No. I guess we could put tables from `attachTags.ejs` and `editTags.ejs` into a single screen. I haven't tried it. Is that what you have in mind?

xjedam commented on views/pages/editTags.ejs:
```diff
@@ -0,0 +1,123 @@
+        <ol class="breadcrumb" style="margin-top: 10px;">
+            <li><a href="/">Home</a></li>
+            <li><a href="/project/<%= project.id %>"><%= project.name %></a></li>
+            <li class="active">Tags</li>
```
I feel like this screen could be the "attach" screen (not the attachTags.ejs) since you could attach or detach tags here. The previous screen feels like Tags edit screen with an odd attach to project functionality glued in. Any particular reason it is done this way? Currently if someone wants to add new tags, or edit existing ones he/she needs to navigate to a project specific tag attachment screen.

Imho this could be `Home/<project name>/Tags` edition screen with a filter to toggle visibility of unattached tags and the previous one could be broader `Home/Tags` edition screen (not linked to any particular project). What do you think?

xjedam commented on views/pages/projectRadar.ejs:
```diff
@@ -44,6 +44,7 @@
                 <li><a href="/project/<%= project.id %>/technology/add">Add Technology</a></li>
                 <li><a href="/project/<%= project.id %>/technology/remove">Remove Technology</a></li>
                 <li><a id="assign-versions-btn" href="#" onclick="reassignVersions()">Reassign versions</a></li>
+                <li><a id="assign-versions-btn" href="/project/<%= project.id %>/tags">Tags</a></li>
```
duplicate `li` tag id

adrianz-kainos replied:
Removed, thx.

xjedam reviewed (commented):
No unit tests ;)?

xjedam commented on handlers/api/tagsApiHandler.js:
```diff
@@ -0,0 +1,92 @@
+
+        var errors = req.validationErrors();
+        if (errors) {
+            res.end(JSON.stringify({success: false, error: errors}));
```
is there some kind of default head written? Don't see res.writeHead here (and a few other instances in this file)

adrianz-kainos replied:
I replaced it with `apiutils.handleResulSet` just like the rest.

adrianz-kainos committed:
Add a shared validator for tags

adrianz-kainos committed:
Refactor attachToProject() in tags DAO,
Use dbhelper in Tag.getAll

adrianz-kainos committed:
Refactor code
Remove redundant script tag
Use apiutils
Remove duplicate element id from projectRadar

adrianz-kainos committed:
Use a single screen to reassign tags.
Edit tags on a separate page.

adrianz-kainos committed:
Use a bootstrap modal to add new tags
Add Google Analytics tracking

adrianz-kainos committed:
Use a bootstrap modal to edit tags

adrianz-kainos committed:
Merge branch 'master' into feature/tags-for-projects

* master:
  change node and npm version to support newest sub-dependency versions
  Fix avatar not loading for adfs users on page with comments
  Switch to .js file instead of .ejs
  Redirect to the index page instead of rendering it
  Fix incorrect attribute name used in validation
  Prevent signing out from the IDP
  Add sticky page size to the remaining tables
  Use cookies to remember the chosen page size
  Add js-cookie script
  Make the technology name column unique
  Add missing user object to server response

adrianz-kainos committed:
Add unit tests for tags API handler
Remove unnecessary (app) argument in tags API handler
Fix typo in passport logger
Add sticky options to the tags table

adrianz-kainos committed:
Merge branch 'master' into feature/tags-for-projects

* master:
  fix connection pooling for pg version 6 on node 6
  fix pg version to the newest one

adrianz-kainos committed:
Add missing error handler in projects DAO
